zookeeper-api/src/main/java/org/apache/helix/zookeeper/api/client/RealmAwareZkClient.java

```diff
@@ -306,48 +326,63 @@ public int hashCode() {
   }
 
   /**
-   * Configuration for creating a new ZkConnection.
+   * ZkConnection-related configs for creating an instance of RealmAwareZkClient.
    */
-  class ZkConnectionConfig {
-    // Connection configs
-    private final String _zkServers;
-    private int _sessionTimeout = HelixZkClient.DEFAULT_SESSION_TIMEOUT;
+  class RealmAwareZkConnectionConfig {
+    /**
+     * mode: Which mode the RealmAwareZkClientConfig should be created in
+     */
+    private final MODE _mode;
+    /**
+     * zkRealmShardingKey: used to deduce which ZK realm this RealmAwareZkClientConfig should connect to.
+     * NOTE: this field will be ignored if MODE is MULTI_REALM!
+     */
+    private final String _zkRealmShardingKey;
+    private int _sessionTimeout = DEFAULT_SESSION_TIMEOUT;
 
-    public ZkConnectionConfig(String zkServers) {
-      _zkServers = zkServers;
+    public RealmAwareZkConnectionConfig(MODE mode, String zkRealmShardingKey) {
```

Comment:
If we think MODE is already encoded in the Type (class hierarchy), we don't need this in the constructor.  Based on looking at the psuedo code I drew, I don't fee we really need to pass this in.

Reply:
I am thinking that we would use this for FederatedZkClient as well since FederatedZkClient is an implementation of RealmAwareZkClient. With that said, this is necessary because the mode will have to be different.

So mode is not necessarily encoded in the type.

Reply:
FederatedZkClient is always multi. The implementation of FederateZkClient can assume this without you passing a mode. The problem is that if you pass a non-multi, it need to handle it then.